### PR TITLE
Custom error message for draw_path. issues : #8131 (bad error message from pyplot.plot)

### PR DESCRIPTION
**Pull request by herilalaina**
Addressing the issue #8131, this PR adds reference to the agg.path.chunksize params in the overflow error message.

### CODE REVIEW

**Comment by NelleV** [comment]
Hi @herilalaina 
Sorry it took so long for us to review the pull request and thanks for the patch!

I'd rather have the try except surround only the part of the code that can raise an overflow error for readabiblity. It means having two try except, but I think we can live with that.
Do you mind making this small change?

**Comment by herilalaina** [comment]
Hi @NelleV, no problem. I will make this change as soon as I can :)

File: lib/matplotlib/backends/backend_agg.py [hunk 1]
```diff
@@ -147,24 +147,34 @@ def draw_path(self, gc, path, transform, rgbFace=None):
...
-                self._renderer.draw_path(gc, p, transform, rgbFace)
+                try:
+                    self._renderer.draw_path(gc, p, transform, rgbFace)
+                except OverflowError as e:
+                    e.args = ("Exceeded cell block limit (set 'agg.path.chunksize' rcparam)",)
+                    raise
         else:
-            self._renderer.draw_path(gc, path, transform, rgbFace)
+            try:
+                self._renderer.draw_path(gc, p, transform, rgbFace)
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_agg.py]: Should be `path`, not `p`.

File: lib/matplotlib/backends/backend_agg.py [hunk 2]
```diff
@@ -147,24 +147,34 @@ def draw_path(self, gc, path, transform, rgbFace=None):
...
                 if c is not None:
                     c = c[ii0:ii1]
-                    c[0] = Path.MOVETO # move to end of last chunk
+                    c[0] = Path.MOVETO  # move to end of last chunk
                 p = Path(v, c)
-                self._renderer.draw_path(gc, p, transform, rgbFace)
+                try:
+                    self._renderer.draw_path(gc, p, transform, rgbFace)
+                except OverflowError as e:
+                    e.args = ("Exceeded cell block limit (set 'agg.path.chunksize' rcparam)",)
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_agg.py]: I'm not sure about this; should we raise a new exception?

File: lib/matplotlib/backends/backend_agg.py (hunk 1 above)
Comment (herilalaina) [inline lib/matplotlib/backends/backend_agg.py]: Sorry about that. thank you

File: lib/matplotlib/backends/backend_agg.py (hunk 2 above)
Comment (herilalaina) [inline lib/matplotlib/backends/backend_agg.py]: I am not sure I understand your point. In fact, the same overflow error occurs if we set `agg.path.chunksize` with a big number. Here, I just change the overflow error message

Comment (QuLogic) [inline lib/matplotlib/backends/backend_agg.py]: I mean, is changing `e.args` a good thing to do, vs raising a new exception entirely.

Comment (herilalaina) [inline lib/matplotlib/backends/backend_agg.py]: in fact, if we raise new overflow error. We get something like this 

> Exception in Tkinter callback
> Traceback (most recent call last):
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/backends/backend_agg.py", line 171, in draw_path
>     self._renderer.draw_path(gc, path, transform, rgbFace)
> OverflowError: In draw_path: Exceeded cell block limit
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib_env/lib/python3.5/tkinter/__init__.py", line 1562, in __call__
>     return self.func(*args)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/backends/backend_tkagg.py", line 284, in resize
>     self.show()
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/backends/backend_tkagg.py", line 355, in draw
>     FigureCanvasAgg.draw(self)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/backends/backend_agg.py", line 475, in draw
>     self.figure.draw(self.renderer)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/artist.py", line 68, in draw_wrapper
>     return draw(artist, renderer, *args, **kwargs)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/figure.py", line 1240, in draw
>     renderer, self, artists, self.suppressComposite)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/image.py", line 139, in _draw_list_compositing_images
>     a.draw(renderer)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/artist.py", line 68, in draw_wrapper
>     return draw(artist, renderer, *args, **kwargs)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/axes/_base.py", line 2386, in draw
>     mimage._draw_list_compositing_images(renderer, self, artists)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/image.py", line 139, in _draw_list_compositing_images
>     a.draw(renderer)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/artist.py", line 68, in draw_wrapper
>     return draw(artist, renderer, *args, **kwargs)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/lines.py", line 800, in draw
>     line_func(renderer, gc, tpath, affine.frozen())
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/lines.py", line 1248, in _draw_solid
>     renderer.draw_path(gc, path, trans)
>   File "/home/herilalaina/Documents/contribution/matplotlib_contrib/matplotlib/lib/matplotlib/backends/backend_agg.py", line 173, in draw_path
>     raise OverflowError("Exceeded cell block limit (set 'agg.path.chunksize' rcparam)")
> OverflowError: Exceeded cell block limit (set 'agg.path.chunksize' rcparam)

I will make the change if It's okay.

Comment (QuLogic) [inline lib/matplotlib/backends/backend_agg.py]: Obviously, it does appear to work, but I don't have an answer whether it is a good or idiomatic thing to be doing or not.

Opinions, @tacaswell?

Comment (herilalaina) [inline lib/matplotlib/backends/backend_agg.py]: Hi @QuLogic , checks failed on coverage. I'm new to code coverage. Do I need to implement test (for overflow error) to fix that ?

File: lib/matplotlib/backends/backend_agg.py [hunk 3]
```diff
@@ -145,8 +144,9 @@ def draw_path(self, gc, path, transform, rgbFace=None):
         """
         Draw the path
         """
-        nmax = rcParams['agg.path.chunksize'] # here at least for testing
+        nmax = rcParams['agg.path.chunksize']  # here at least for testing
         npts = path.vertices.shape[0]
+
         if (nmax > 100 and npts > nmax and path.should_simplify and
```
Comment (phobson) [inline lib/matplotlib/backends/backend_agg.py]: This half of the `if/else` block is untested. Think you could add a test?

Comment (herilalaina) [inline lib/matplotlib/backends/backend_agg.py]: hi @phobson , thank you. I will do it

**Comment by tacaswell** [comment]
:+1: on this.  To test the missed branch I assume you need to user a _very_ long line with a chunk size that is too big?

File: lib/matplotlib/tests/test_agg.py [hunk 4]
```diff
@@ -219,3 +219,23 @@ def test_too_large_image():
...
+    tmp = rcParams['agg.path.chunksize']
+    rcParams['agg.path.chunksize'] = 500
+    ax.plot(x, np.sin(x))
+    plt.show()
+
+    # Test with big chunksize
+    rcParams['agg.path.chunksize'] = 8000000
+    with pytest.raises(OverflowError):
+        ax.plot(x, np.sin(x))
+        plt.show()
```
Comment (herilalaina) [inline lib/matplotlib/tests/test_agg.py]: Hi @tacaswell , it doesn't raise OverflowError. Do you have any idea how to handle it?

Comment (tacaswell) [inline lib/matplotlib/tests/test_agg.py]: Is that chunk size big enough to cause a overflow?

Comment (herilalaina) [inline lib/matplotlib/tests/test_agg.py]: It never raises overflow even if I increase chunk size. The test just crashed

File: lib/matplotlib/tests/test_agg.py [hunk 5]
```diff
@@ -219,3 +219,17 @@ def test_too_large_image():
...
         fig.savefig(buff)
+
+
+def test_chunksize():
+    x = range(9000000)
+    fig = plt.figure()
+    ax = fig.add_subplot(1, 1, 1)
+
+    # Test with small chunksize
+    tmp = rcParams['agg.path.chunksize']
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_agg.py]: You don't need to worry about this; the cleanup on the test will reset all `rcParams`.

File: lib/matplotlib/tests/test_agg.py [hunk 6]
```diff
@@ -219,3 +219,17 @@ def test_too_large_image():
     buff = io.BytesIO()
     with pytest.raises(ValueError):
         fig.savefig(buff)
+
+
+def test_chunksize():
+    x = range(9000000)
+    fig = plt.figure()
+    ax = fig.add_subplot(1, 1, 1)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_agg.py]: `fig, ax = plt.subplots()`

**Comment by QuLogic** [comment]
I think the test is not doing anything because of the `plt.show`; replacing it with `fig.canvas.draw()` at least triggers something, though it still doesn't seem to cause an exception...

**Comment by QuLogic** [comment]
Actually, if all you want to do is test that side of the branch, you only need a path longer than the chunksize (which must be greater than 100), so set the rcParam to 105 and make a path that's 110 long. 9000000 is much too long, I think.

If you want to _trigger the exception_, that might be a little more difficult and I'm not sure how to do that.

**Comment by QuLogic** [comment]
After some testing, it looks like you need roughly `range(200000)` to trigger the exception with `chunksize<=100`. That means that you'd need approximately `range(20000000)` to trigger an exception with `chunksize>100`. I think this is basically impossible so I don't think we should try to trigger the exception in that case.

**Comment by herilalaina** [comment]
Thank you for your great feedback ! I will fix them.

**Pull request merged**